[MachineOutliner][NFC] Refactor

kyulee-com opened the pull request:
This patch prepares the NFC groundwork for global outlining using CGData, which will follow https://github.com/llvm/llvm-project/pull/90074.

 - The `MinRepeats` parameter is now explicitly passed to the `getOutliningCandidateInfo` function, rather than relying on a default value of 2. For local outlining, the minimum number of repetitions is typically 2, but for the global outlining (mentioned above), we will optimistically create a single `Candidate` for each `OutlinedFunction` if stable hashes match a specific code sequence. This parameter is adjusted accordingly in global outlining scenarios.
 - I have also implemented `unique_ptr` for `OutlinedFunction` to ensure safe and efficient memory management within `FunctionList`, avoiding unnecessary implicit copies.

This depends on https://github.com/llvm/llvm-project/pull/101461.
This is a patch for https://discourse.llvm.org/t/rfc-enhanced-machine-outliner-part-2-thinlto-nolto/78753.

kyulee-com committed:
[MachineOutliner][NFC] Refactor

ellishg commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -2088,14 +2088,22 @@ class TargetInstrInfo : public MCInstrInfo {
       const MachineModuleInfo &MMI,
-      std::vector<outliner::Candidate> &RepeatedSequenceLocs) const {
+      std::vector<outliner::Candidate> &RepeatedSequenceLocs,
+      unsigned MipRep) const {
```
```suggestion
      unsigned MinRep) const {
```

Since characters are cheap, can we expand this like `MinRepeates`? Or maybe `MinSequenceCount`?

ellishg replied:
Also, can we do `unsigned MinRep=2` to set a default value instead of creating a new function?

kyulee-com replied:
I just made `MinRepeates` explicit for `getOutliningCandidateInfo`. I don't want it to be a field/property as TargetInstrInfo as this value can be overridden by the client (either for creating candidates for the local outliner vs. for the global outliner).

ellishg commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -684,7 +688,7 @@ void MachineOutliner::findCandidates(
     }
 
-    FunctionList.push_back(*OF);
+    FunctionList.push_back(std::make_unique<OutlinedFunction>(*OF));
```
Does this work?
```suggestion
    FunctionList.emplace_back(*OF);
```

kyulee-com replied:
This isn't quite right, but either case created implicit copy from the local OutlinedFunction instance. Instead, I also make it unique when it's created in `getOutliningCandidateInfo` so that we can simply move it here.

aemerson commented:
Can you elaborate more in the PR/commit message about why you're making some of these changes? Things like the 2 -> MinRep. Taken as an individual change it would be good to have a rational without depending on future changes to make sense of it.

kyulee-com committed:
Address comments from ellishg

kyulee-com commented:
> Can you elaborate more in the PR/commit message about why you're making some of these changes? Things like the 2 -> MinRep. Taken as an individual change it would be good to have a rational without depending on future changes to make sense of it.

Yeah. I updated the summary with some details + the link for the next candidate PR.

kyulee-com commented:
Can someone take a look at this NFC (hopefully straightforward) code?

michaelmaitland commented:
> The cost functions within the OutlinedFunction class have been made virtual, allowing for future overrides by GlobalOutlinedFunction. This is crucial because GlobalOutlinedFunction does not have access to Candidates across modules, which represent repeated sequences in the current module. Instead, it uses a proxy of global candidate counts for cost computation.

It sounds like there is an assumption that there will be future overrides by GlobalOutlinedFunction, but it isn't clear when or even if this will ever happen. Maybe it would be better to do this in the same patch where someone is trying to override GlobalOutlinedFunction.

michaelmaitland reviewed (commented):
In multiple places there is typo: `MinRepeates` -> `MinRepeats`?

kyulee-com committed:
Address comments from michaelmaitland

kyulee-com commented:
> In multiple places there is typo: `MinRepeates` -> `MinRepeats`?

Thanks for the catch! I blindly copied & pasted it from the prior suggestion.

kyulee-com committed:
Address comments from thevinster

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -670,21 +674,24 @@ void MachineOutliner::findCandidates(
 
-    std::optional<OutlinedFunction> OF =
-        TII->getOutliningCandidateInfo(*MMI, CandidatesForRepeatedSeq);
+    unsigned MinRepeats = 2;
```
I hoisted it. I keep using the variable instead of a hard-coded value for better readability. I believe the efficiency-wise, it's the same as the compiler will optimize it.

thevinster reviewed (approved):
lgtm

kyulee-com commented:
Thank you everyone for your comments! I believe I have addressed all the comments. If there's anything else that needs attention, I will take care of it in the next PR.
